staging/src/k8s.io/client-go/util/homedir/homedir.go

```diff
@@ -21,27 +21,49 @@ import (
 	"runtime"
 )
 
-// HomeDir returns the home directory for the current user
+// HomeDir returns the home directory for the current user.
 func HomeDir() string {
+	return PreferredHomeDir(nil)
+}
+
+// PreferredHomeDir returns the home directory for the current user.
+// If there are multiple possible home directories (as on windows), all possible paths
+// are tested with each path test in order, and the first path to pass a test is returned.
+// If no paths pass any tests, or no path tests are provided, the first possible path is returned.
+func PreferredHomeDir(pathTests []func(string) bool) string {
 	if runtime.GOOS == "windows" {
 
+		possiblePaths := []string{}
 		// First prefer the HOME environmental variable
 		if home := os.Getenv("HOME"); len(home) > 0 {
 			if _, err := os.Stat(home); err == nil {
-				return home
+				possiblePaths = append(possiblePaths, home)
 			}
 		}
 		if homeDrive, homePath := os.Getenv("HOMEDRIVE"), os.Getenv("HOMEPATH"); len(homeDrive) > 0 && len(homePath) > 0 {
 			homeDir := homeDrive + homePath
 			if _, err := os.Stat(homeDir); err == nil {
-				return homeDir
+				possiblePaths = append(possiblePaths, homeDir)
 			}
 		}
 		if userProfile := os.Getenv("USERPROFILE"); len(userProfile) > 0 {
 			if _, err := os.Stat(userProfile); err == nil {
-				return userProfile
+				possiblePaths = append(possiblePaths, userProfile)
+			}
+		}
+
+		// Return the first possible path that passes the first provided test
+		for _, pathTest := range pathTests {
```

Comment:
@mbrancato
> If `HOMEDRIVE/HOMEPATH` is writable it should still be ignored if it does not have a `.kube` folder

how would you evaluate the alternative?
```
for _, possiblePath := range possiblePaths {
	for _, pathTest := range pathTests {
		if pathTest(possiblePath) {
			return possiblePath
		}
	}
}
```

Reply:
Sorry, I'm not sure what you're asking of me here, but I think it may need to check all paths before returning instead of returning the first match.

Reply:
sorry, i wrote that after a long day.